### PR TITLE
NIFI-10344: refactored groovy tests in nifi-standard-processor to jav…

### PR DESCRIPTION
**Pull request by emiliosetiadarma**
…a and JUnit 5

# Summary

[NIFI-10344](https://issues.apache.org/jira/browse/NIFI-10344)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+public class CountTextTest {
+    private static final String TLC = "text.line.count";
+    private static final String TLNEC = "text.line.nonempty.count";
+    private static final String TWC = "text.word.count";
+    private static final String TCC = "text.character.count";
+
+    @BeforeAll
+    static void setUpOnce() {
+        Security.addProvider(new BouncyCastleProvider());
+    }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java:53]: This Security Provider configuration method is not necessary and should be removed.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+    @Test
+    void testShouldCountAllMetrics() throws IOException {
+        final TestRunner runner = TestRunners.newTestRunner(CountText.class);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java:57]: This line could be moved to a shared `setRunner` method annotated with `BeforeEach` to create the runner.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        runner.run();
+
+        runner.assertAllFlowFilesTransferred(CountText.REL_SUCCESS, 1);
+        FlowFile flowFile = runner.getFlowFilesForRelationship(CountText.REL_SUCCESS).get(0);
+        for (final Map.Entry<String, String> entry: expectedValues.entrySet()) {
+            final String attribute = entry.getKey();
+            final String expectedValue = entry.getValue();
+
+            assertEquals(expectedValue, flowFile.getAttribute(attribute));
+        }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java:83]: This approach should be replaced with `MockFlowFile.assertAttributeEquals()`, which would also remove the need for the `Map` of expected values.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+    @BeforeAll
+    static void setUpOnce() {
+        Security.addProvider(new BouncyCastleProvider());
+    }
+
+    @Test
+    void testShouldCalculateHashOfPresentAttribute() {
+        final HashAlgorithm[] algorithms = HashAlgorithm.values();
+
+        final TestRunner runner = TestRunners.newTestRunner(new CryptographicHashAttribute());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:51]: This should be moved to a `setRunner()` method annotated with `BeforeEach`.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+public class CryptographicHashAttributeTest {
+    @BeforeAll
+    static void setUpOnce() {
+        Security.addProvider(new BouncyCastleProvider());
+    }
+
+    @Test
+    void testShouldCalculateHashOfPresentAttribute() {
+        final HashAlgorithm[] algorithms = HashAlgorithm.values();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:49]: Declaring this local variable is unnecessary.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+        // Create attributes for username and date
+
+        final Map<String, String> attributes = new HashMap<>();
+        attributes.put("username", "alopresto");
+        attributes.put("date", ZonedDateTime.now().format(DateTimeFormatter.ofPattern("YYYY-MM-dd HH:mm:ss.SSS Z")));
+
+        final Set<String> attributeKeys = attributes.keySet();
+
+        for (final HashAlgorithm algorithm : algorithms) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:61]:
```suggestion
        for (final HashAlgorithm algorithm : HashAlgorithm.values()) {
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+            final String EXPECTED_USERNAME_HASH = HashService.hashValue(algorithm, attributes.get("username"));
+            final String EXPECTED_DATE_HASH = HashService.hashValue(algorithm, attributes.get("date"));
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:63]: These variable names should be lowercased following standard conventions for local variables.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    static void setUpOnce() {
+        Security.addProvider(new BouncyCastleProvider());
+    }
+
+    @Test
+    void testShouldCalculateHashOfPresentContent() throws IOException {
+        final HashAlgorithm[] algorithms = HashAlgorithm.values();
+
+        // Generate some long content (90 KB)
+        final String LONG_CONTENT = StringUtils.repeat("apachenifi ", 8192);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java:51]: This variable should be lowercased to follow standard conventions.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+        final TestRunner runner = TestRunners.newTestRunner(new CryptographicHashContent());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java:53]: Should be moved to a standard `BeforeEach` annotated method.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/SplitXmlTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    @Test
+    void testShouldHandleXXEInTemplate() throws IOException {
+        final String XXE_TEMPLATE_FILEPATH = "src/test/resources/xxe_template.xml";
+        final TestRunner runner = TestRunners.newTestRunner(new SplitXml());
+        runner.setProperty(SplitXml.SPLIT_DEPTH, "3");
+        runner.enqueue(Paths.get(XXE_TEMPLATE_FILEPATH));
+
+        runner.run();
+
+        runner.assertAllFlowFilesTransferred(SplitXml.REL_FAILURE);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/SplitXmlTest.java:36]: Recommend moving this shared code to a new private method named something like `assertExternalEntitiesFailure()` and changing both test methods to call the shared method.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFlattenJson.java [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+            parsed = mapper.readValue(asJson, List.class);
+            assertEquals(keyCount, parsed.size(), "Too many keys");
+            return parsed;
+        } else {
+            final Map parsed;
+            parsed = mapper.readValue(asJson, Map.class);
+            assertEquals(keyCount, parsed.size(), "Too many keys");
+            return parsed;
+        }
+    }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFlattenJson.java:89]: All private methods should be moved to the end of the class, following standard conventions.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for putting in the work to refactor these tests @emiliosetiadarma! Having these tests in Java and JUnit 5 will make them more maintainable going forward.

The majority of the changes look good. I noted a few places where variable names could be changed and some common code could be moved to a shared method. In most cases, I highlighted one example, as opposed to noting the same thing for every occurrence of the issue in a particular test class.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java (hunk 1 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java:53]: Removing it!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java (hunk 2 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java:57]: Making the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java (hunk 3 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CountTextTest.java:83]: Making the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java (hunk 4 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:51]: Making the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java (hunk 5 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:49]: Making the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java (hunk 6 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:61]: Making the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java (hunk 7 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashAttributeTest.java:63]: Made the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java (hunk 8 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java:51]: Lowercased in addition to the other variables defined in the test

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java (hunk 9 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/CryptographicHashContentTest.java:53]: Made the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/SplitXmlTest.java (hunk 10 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/SplitXmlTest.java:36]: Made the changes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFlattenJson.java (hunk 11 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFlattenJson.java:89]: Moved it!

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the feedback and refactoring these tests @emiliosetiadarma, the latest version looks good! +1 merging

**Pull request closed without merging**